src/System.Management.Automation/engine/CommandProcessorBase.cs

```diff
@@ -937,21 +948,77 @@ public void Dispose()
         private void Dispose(bool disposing)
         {
             if (_disposed)
+            {
                 return;
+            }
 
-            if (disposing)
+            try
             {
-                // 2004/03/05-JonN Look into using metadata to check
-                // whether IDisposable is implemented, in order to avoid
-                // this expensive reflection cast.
-                IDisposable id = Command as IDisposable;
-                if (id != null)
+                if (disposing)
                 {
-                    id.Dispose();
+                    Pipe oldErrorOutputPipe = _context.ShellFunctionErrorOutputPipe;
+                    CommandProcessorBase oldCurrentCommandProcessor = _context.CurrentCommandProcessor;
+
+                    bool isStopping = false;
+                    try
+                    {
+                        if (this.RedirectShellErrorOutputPipe || _context.ShellFunctionErrorOutputPipe != null)
+                        {
+                            _context.ShellFunctionErrorOutputPipe = this.commandRuntime.ErrorOutputPipe;
+                        }
+
+                        _context.CurrentCommandProcessor = this;
+                        isStopping = ExceptionHandlingOps.SuspendStoppingPipeline(Context);
+                        SetCurrentScopeToExecutionScope();
+
+                        // This is a no-op for compiled cmdlets
+                        CleanupScriptCommands();
```

Comment:
Hmmm, good point. I can probably move some of this into the `DisposeScriptCommands` method itself.

Reply:
I think I was trying to avoid excessive branching, and had some trouble figuring out how to both keep the try/finally structure needed for the scope method calls and also avoid needing to double up on the `if (processor is DlrScriptCommandProcessor)` / `if (processor.Command is PSScriptCmdlet)` checks to skip it for the compiled commands.

Reply:
I can move... probably a lot of this into `DisposeScriptCommands`, I think. That would be a ton cleaner. Let me see how that works out.

Reply:
pulled the majority of this into `DisposeScriptCommand` 🙂